src/libs/sys-util-rs/src/cgroup.rs

```diff
@@ -0,0 +1,690 @@
+// Copyright (c) 2019-2021 Alibaba Cloud
+// Copyright (c) 2019-2021 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::ops::Deref;
+use std::path::{Component, Path, PathBuf};
+use std::sync::Mutex;
+
+use cgroups::{Cgroup, CgroupPid, Controllers, Hierarchy, Subsystem};
+use lazy_static::lazy_static;
+use once_cell::sync::Lazy;
+
+use crate::sl;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Can not add tgid {0} to cgroup, {1:?}")]
+    AddTgid(u64, #[source] cgroups::error::Error),
+    #[error("failed to apply resources to cgroup: {0:?}")]
+    ApplyResource(#[source] cgroups::error::Error),
+    #[error("failed to delete cgroup after {0} retries")]
+    DeleteCgroup(u64),
+    #[error("Invalid cgroup path {0}")]
+    InvalidCgroupPath(String),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+lazy_static! {
+    /// Disable cgroup v1 subsystems.
+    pub static ref DISABLED_HIERARCHIES: Mutex<Vec<String>> = Mutex::new(Vec::new());
+}
+
+/// Update the disabled cgroup subsystems.
+pub fn update_disabled_cgroup_list(hierarchies: &[String]) {
+    let mut disabled_hierarchies = DISABLED_HIERARCHIES.lock().unwrap();
+    disabled_hierarchies.clear();
+    for hie in hierarchies {
+        disabled_hierarchies.push(hie.clone());
+    }
+    info!(
+        sl!(),
+        "disable cgroup list {:?} from {:?}", disabled_hierarchies, hierarchies
+    );
+}
+
+/// Filter out disabled cgroup subsystems.
+pub fn filter_disabled_cgroup(controllers: &mut Vec<Controllers>) {
+    let disabled_hierarchies = DISABLED_HIERARCHIES.lock().unwrap();
+    let mut blacklist_controllers: Vec<Controllers> = vec![];
```

Comment:
Why not parse the string names into `Controllers` values in `update_disabled_cgroup_list` so you only have to keep around a vector of simple enums rather than a bulky vector of strings?

Reply:
Good suggestion!